crates/matrix-sdk/src/test_utils/mocks.rs

```diff
@@ -0,0 +1,389 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(missing_debug_implementations)]
+
+use std::sync::{Arc, Mutex};
+
+use matrix_sdk_base::deserialized_responses::TimelineEvent;
+use matrix_sdk_test::{
+    test_json, InvitedRoomBuilder, JoinedRoomBuilder, KnockedRoomBuilder, LeftRoomBuilder,
+    SyncResponseBuilder,
+};
+use ruma::{OwnedEventId, OwnedRoomId, RoomId};
+use serde_json::json;
+use wiremock::{
+    matchers::{header, method, path, path_regex},
+    Mock, MockBuilder, MockGuard, MockServer, Respond, ResponseTemplate, Times,
+};
+
+use super::logged_in_client;
+use crate::{Client, Room};
+
+pub struct MatrixMockServer {
+    server: MockServer,
+    client: Client,
```

Comment:
Passing the client upon construction of the mock server is a prerequisite, so the server's URI is injected into the client, at construction; reverting this would make construction a bit unwieldy, as one would have to pass a `ClientBuilder`, or get the URL from the `MatrixMockServer` and use it to create a `Client`, leading to more duplicated code for all callers.

Happy to rename this struct overall to something else, though? Not sure what a good name would be…